library/nist_kw.c

```diff
@@ -0,0 +1,649 @@
+    if( cipher_info == NULL )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    if( cipher_info->block_size != 16 )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    mbedtls_cipher_free( &ctx->cipher_ctx );
+
+    if( ( ret = mbedtls_cipher_setup( &ctx->cipher_ctx, cipher_info ) ) != 0 )
+        return( ret );
+
+    if( ( ret = mbedtls_cipher_setkey( &ctx->cipher_ctx, key, keybits,
+                               is_wrap ? MBEDTLS_ENCRYPT : MBEDTLS_DECRYPT ) ) != 0 )
+    {
+        return( ret );
+    }
+
+    return( 0 );
+}
+
+/*
+ * Free context
+ */
+void mbedtls_nist_kw_free( mbedtls_nist_kw_context *ctx )
+{
+    mbedtls_cipher_free( &ctx->cipher_ctx );
+    mbedtls_platform_zeroize( ctx, sizeof( mbedtls_nist_kw_context ) );
+}
+
+/*
+ * KW-AE as defined in SP 800-38F section 6.2
+ * KWP-AE as defined in SP 800-38F section 6.3
+ */
+int mbedtls_nist_kw_wrap( mbedtls_nist_kw_context *ctx, mbedtls_nist_kw_mode_t mode,
+                          const unsigned char *input, size_t in_len,
+                          unsigned char *output, size_t* out_len )
+{
+    int ret = 0;
+    size_t semiblocks = 0;
+    size_t s ;
+    size_t olen, i, padlen = 0;
+    uint64_t t = 0;
+    unsigned char outbuff[KW_SEMIBLOCK_LENGTH * 2];
+    unsigned char inbuff[KW_SEMIBLOCK_LENGTH * 2];
+    unsigned char* R2 = output + KW_SEMIBLOCK_LENGTH;
+    unsigned char* A = output ;
+    const uint32_t ICV1[] = NIST_KW_ICV1;
+    const uint32_t ICV2[] = NIST_KW_ICV2;
+
+    /*
+     * Generate the String to work on
+     */
+    if( mode == MBEDTLS_KW_MODE_KW )
+    {
+        /*
+         * According to SP800-38F Table 1, the plaintext length for KW
+         * must be between 2 to 2^54-1 semiblocks inclusive.
+         */
+        if( ( in_len < 8 )
+#if SIZE_MAX > UINT_MAX
+                || ( in_len > 0x1FFFFFFFFFFFFF8 )
+#endif
+           )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+
+
+        memcpy( output, ICV1, KW_SEMIBLOCK_LENGTH );
+        memcpy( output + KW_SEMIBLOCK_LENGTH, input, in_len );
+    }
+    else
+    {
+        /*
+         * According to SP 800-38F Table 1, the plaintext length for KWP
+         * must be between 1 and 2^32-1 octets inclusive.
+         */
+        if( ( in_len < 1 )
+#if SIZE_MAX > UINT_MAX
+           || ( in_len > 0xFFFFFFFF )
+#endif
+          )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+        }
+
+        if( ( in_len % 8 ) != 0 )
+        {
+            padlen = ( 8 - ( in_len % 8 ) );
+        }
+
+        memcpy( output, &ICV2, KW_SEMIBLOCK_LENGTH / 2 );
+        PUT_UINT32_BE( ( in_len & 0xffffffff ), output, KW_SEMIBLOCK_LENGTH / 2 );
+
+        memcpy( output + ( KW_SEMIBLOCK_LENGTH / 2 ) + sizeof(uint32_t), input, in_len );
+        memset( output + ( KW_SEMIBLOCK_LENGTH / 2 ) + sizeof(uint32_t) + in_len, 0, padlen );
```

Comment:
style: `sizeof ( uint32_t )`, but I don't understand what's going on here. Why is 4 expressed as both `KW_SEMIBLOCK_LENGTH / 2` and `sizeof( uint32_t )` inside the same sum?